### PR TITLE
add firewaller support to Equinix Metal provider

### PR DESCRIPTION
**Pull request by displague**
Updates the Equinix Metal provider to support Firewaller via SSH+IPTables (for `juju export` / `juju unexport`)
* implements `environs.Firewaller` 
* implements `instances.InstanceFirewaller`

Also:
* implements `environs.InstanceTagger`

Continues #13552 

## QA steps

```sh
git clone https://github.com/juju/juju/
cd juju
git fetch origin pulls/13695/head:equinixmetal-firewaller
git checkout equinixmetal-firewaller
go install ./cmd/juju
which juju # should be the one installed in the GOBIN path

YOUR_IP=$(curl -4 icanhazip.com)
YOUR_IPV6=$(curl -6 icanhazip.com)
INSTANCE_TYPE=c3.medium.x86 # pick the smallest plan available to you in Dallas

juju add-credential equinix
juju bootstrap equinix/da equinix-controller \
  --bootstrap-image=ubuntu_20_04 \
  --bootstrap-series=focal \
  --bootstrap-constraints="arch=amd64 instance-type=$INSTANCE_TYPE" \
  --build-agent # to use the local linux environment development build remotely

juju set-model-constraints arch=amd64 instance-type=$INSTANCE_TYPE

juju deploy mysql --channel=edge --series=bionic && \
juju deploy wordpress --channel=edge --series=bionic --to=0 && \
juju add-relation wordpress mysql && \
juju expose wordpress --to-cidrs=${YOUR_IP}/32 # your ipv4, add ",${YOUR_IPV6}/128" if also available
# ^ it should now be accessible
# juju status to get the IP and load it in your browser

juju unexpose wordpress
# ^ it should no longer be accessible. verify in your browser.

# debug: if the page loads when it is not exposed, report the following log:
# TODO: how to log helpful debuggin
```

```
# cleanup
juju destroy-controller equinix-controller  --destroy-all-models
```

Next: TODO

## Bug reference
https://bugs.launchpad.net/juju/+bug/1940520

### CODE REVIEW

**Comment by achilleasa** [comment]
On a _quick_ glance, the logic seems correct. 

However what is missing is some extra cloud-init scripts to lock down access to the machines that Juju provisions. In particular, for **non-controller** instances, you would need to add iptables rules to block INGRESS traffic on the NIC that has been assigned a public IP but allow EGRESS (outgoing, established, and most importantly, **related** sessions).

This should allow the machine to reach the internet so it can download the resources it needs but would otherwise prevent incoming connections unless you `juju expose` an application in which case the firewaller code from this PR would be invoked to update the ingress rules.

**Comment by achilleasa** [comment]
@manadart can you please track this PR? It still needs some work before it can correctly address the issue referenced in the attached bug.

**Comment by achilleasa** [comment]
@displague Can you please update the PR description and add some commands that folks can use as QA steps for the PR? 

For example, a nice scenario for testing would be: bootstrap, deploy  a charm such as redis and verify that it's port is not accessible via the public IP assigned to the box. Then, `juju expose redis` (also try the `--to-cidr` argument to make sure it works) and check that the port becomes accessible. Finally, un-expose the app and verify that once more the port is not accessible.

**Comment by manadart** [comment]
@displague Happy to push this through when Achilleas' comments above are addressed. 

Otherwise, if this is no longer the preferred solution, we can close it.

**Review by manadart — CHANGES_REQUESTED** [review]
I see where this is failing now.

Upon running `juju expose`, this log line can be seen:
```
controller-0: 19:16:06 INFO juju.worker.firewaller flushInstancePorts called on an instance of type *equinix.equinixDevice which doesn't support firewall.
```
You need to to modify the instance to implement `InstanceFirewaller` from _environs/instances/instance.go_.

The instance configurator doesn't work as an environ firewaller. See https://github.com/juju/juju/pull/10288 for an example of adding it to the OCI provider.

**Comment by displague** [comment]
@manadart What commands are you using (or where are you looking) to see that log message?


Also, I tried the `juju deploy '~juju-qa/network-health'` example from OCI PR, but this want not found.

**Comment by manadart** [comment]
Yes, that patch dates from before Charmstore was superseded by Charmhub. You should be able to show it up by exposing any application.

File: provider/equinix/firewaller.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// IngressRules returns the port ranges opened for the whole environment.
+// Must only be used if the environment was setup with the
+// FwGlobal firewall mode.
+func (c *equinixFirewaller) IngressRules(ctx context.ProviderCallContext) (firewall.IngressRules, error) {
+	return nil, errors.NotSupportedf("Ports")
+}
+
+// DeleteGroups implements OpenstackFirewaller interface.
```
Comment (manadart) [inline provider/equinix/firewaller.go:39]: We want to delete from here down in this file.

**Review by manadart — COMMENTED** [review]
Almost there. Some dead code to remove.

**Review by manadart — APPROVED** [review]
QA'd just using an app directly on the controller.

I can see the ingress rules applied:
```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
ACCEPT     tcp  --  <my IP>  anywhere             tcp dpt:x11 /* juju ingress */
```

**Comment by displague** [comment]
@manadart I ran my test steps again on 2.9 and I see what you saw. 

```
$ iptables -L -t filter 
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         
ACCEPT     tcp  --  <my-ip>  anywhere             tcp dpt:http /* juju ingress */
...
```

The default input policy is ACCEPT, however, so with the expose or unexpose, I'm still able to access the service.
Should additional rules be inserted to change the default policy to DROP? Does iptables-persistent need to be installed?

**Pull request merged**